tempodb/encoding/vparquet2/block_traceql.go

```diff
@@ -304,45 +311,49 @@ func (i *bridgeIterator) Next() (*pq.IteratorResult, error) {
 		// The spanset is in the OtherEntries
 		iface := res.OtherValueFromKey(otherEntrySpansetKey)
 		if iface == nil {
-			return nil, fmt.Errorf("engine assumption broken: spanset not found in other entries")
+			return nil, fmt.Errorf("engine assumption broken: spanset not found in other entries in bridge")
 		}
 		spanset, ok := iface.(*traceql.Spanset)
 		if !ok {
-			return nil, fmt.Errorf("engine assumption broken: spanset is not of type *traceql.Spanset")
+			return nil, fmt.Errorf("engine assumption broken: spanset is not of type *traceql.Spanset in bridge")
 		}
 
-		var filteredSpansets []*traceql.Spanset
-		if i.cb != nil {
-			filteredSpansets, err = i.cb(spanset)
-			if err == io.EOF {
-				return nil, nil
-			}
-			if err != nil {
-				return nil, err
-			}
-			// if the filter removed all spansets then let's release all back to the pool
-			// no reason to try anything more nuanced than this. it will handle nearly all cases
-			if len(filteredSpansets) == 0 {
-				for _, s := range spanset.Spans {
-					putSpan(s.(*span))
-				}
+		filteredSpansets, err := i.cb(spanset)
+		if err == io.EOF {
+			return nil, nil
+		}
+		if err != nil {
+			return nil, err
+		}
+		// if the filter removed all spansets then let's release all back to the pool
+		// no reason to try anything more nuanced than this. it will handle nearly all cases
+		if len(filteredSpansets) == 0 {
+			for _, s := range spanset.Spans {
+				putSpan(s.(*span))
 			}
-		} else {
-			filteredSpansets = []*traceql.Spanset{spanset}
 		}
 
 		// flatten spans into i.currentSpans
 		for _, ss := range filteredSpansets {
-			for _, s := range ss.Spans {
+			for idx, s := range ss.Spans {
 				span := s.(*span)
-				i.currentSpans = append(i.currentSpans, span)
+
+				// use otherEntryCallbackSpansetKey to indicate to the rebatchIterator that either
+				// 1) this is the last span in the spanset, or 2) there are more spans in the spanset
+				span.cbSpansetFinal = idx == len(ss.Spans)-1
+				span.cbSpanset = ss
```

Comment:
we could do that, but then the rebatchIterator would have to call Next() until it received a nil span to guarantee it's batches were complete. Then it would be able to dump all batches. I did this to preserve the "streaming" nature of the iterators